File: frontend/src/services/eventService.ts
/**
 * EventService - WebSocket client for real-time event communication
 *
 * Provides a singleton Socket.IO client that connects to the backend
 * WebSocket server for real-time event updates.
 */

import { io, Socket } from "socket.io-client"

// Event types matching backend EventType
export const EventTypes = {
  // Session events
  SESSION_CREATED: "session_created",
  SESSION_UPDATED: "session_updated",
  SESSION_DELETED: "session_deleted",

  // Message events
  MESSAGE_CREATED: "message_created",
  MESSAGE_UPDATED: "message_updated",
  MESSAGE_DELETED: "message_deleted",

  // Activity events
  ACTIVITY_CREATED: "activity_created",
  ACTIVITY_UPDATED: "activity_updated",
  ACTIVITY_DELETED: "activity_deleted",

  // Agent events
  AGENT_CREATED: "agent_created",
  AGENT_UPDATED: "agent_updated",
  AGENT_DELETED: "agent_deleted",

  // Environment events
  ENVIRONMENT_ACTIVATING: "environment_activating",
  ENVIRONMENT_ACTIVATED: "environment_activated",
  ENVIRONMENT_ACTIVATION_FAILED: "environment_activation_failed",
  ENVIRONMENT_SUSPENDED: "environment_suspended",

  // Streaming events
  STREAM_STARTED: "stream_started",
  STREAM_COMPLETED: "stream_completed",
  STREAM_ERROR: "stream_error",

  // To=do progress events (from TodoWrite tool)
  TODO_LIST_UPDATED: "todo_list_updated",    // Session-level to-do update
  TASK_TODO_UPDATED: "task_todo_updated",    // Task-level to-do update

  // Generic notification
  NOTIFICATION: "notification",
} as const

export type EventType = (typeof EventTypes)[keyof typeof EventTypes]

// Event structure matching backend EventPublic
export interface EventData {
  type: string
  model_id?: string
  text_content?: string
  meta?: Record<string, any>
  user_id?: string
  timestamp: string
}

// Event handler callback type
export type EventHandler = (event: EventData) => void

// Event subscription
interface EventSubscription {
  eventType: string | "*" // "*" means subscribe to all events
  handler: EventHandler
  id: string
}

// Connection status
export type ConnectionStatus = "connected" | "connecting" | "disconnected"

// Connection status listener
export type ConnectionStatusListener = (status: ConnectionStatus) => void

class EventServiceClass {
  private socket: Socket | null = null
  private subscriptions: Map<string, EventSubscription> = new Map()
  private activeRooms: Set<string> = new Set()
  private isConnecting = false
  private reconnectAttempts = 0
  private maxReconnectAttempts = 5
  private reconnectDelay = 1000 // Start with 1 second
  private statusListeners: Set<ConnectionStatusListener> = new Set()
  private currentStatus: ConnectionStatus = "disconnected"

  /**
   * Initialize the WebSocket connection
   */
  connect(userId: string): void {
    if (this.socket?.connected || this.isConnecting) {
      console.log("[EventService] Already connected or connecting")
      return
    }

    this.isConnecting = true
    this.updateStatus("connecting")
    const apiUrl = import.meta.env.VITE_API_URL || "http://localhost:8000"

    console.log("[EventService] Connecting to:", apiUrl, "with path: /ws")

    this.socket = io(apiUrl, {
      path: "/ws",
      transports: ["websocket", "polling"],
      auth: {
        user_id: userId,
      },
      reconnection: true,
      reconnectionAttempts: this.maxReconnectAttempts,
      reconnectionDelay: this.reconnectDelay,
      reconnectionDelayMax: 5000,
    })

    // Connection event handlers
    this.socket.on("connect", () => {
      console.log("[EventService] Connected, socket ID:", this.socket?.id)
      this.isConnecting = false
      this.reconnectAttempts = 0
      this.reconnectDelay = 1000
      this.updateStatus("connected")

      // Re-subscribe to any active rooms after reconnect
      if (this.activeRooms.size > 0) {
        this.activeRooms.forEach((room) => {
          console.log("[EventService] Re-subscribing to room:", room)
          this.socket?.emit("subscribe", { room }, (response: any) => {
            if (response?.status === "success") {
              console.log(`[EventService] Re-subscribed to room: ${room}`)
            } else {
              console.error(`[EventService] Failed to re-subscribe to room: ${room}`, response)
            }
          })
        })
      }
    })

    this.socket.on("disconnect", (reason) => {
      console.log("[EventService] Disconnected:", reason)
      this.isConnecting = false
      this.updateStatus("disconnected")
    })

    this.socket.on("connect_error", (error) => {
      console.error("[EventService] Connection error:", error)
      this.isConnecting = false
      this.reconnectAttempts++

      // Exponential backoff
      if (this.reconnectAttempts < this.maxReconnectAttempts) {
        this.reconnectDelay = Math.min(this.reconnectDelay * 2, 5000)
        this.updateStatus("connecting")
      } else {
        this.updateStatus("disconnected")
      }
    })

    // Listen for events from server
    this.socket.on("event", (data: EventData) => {
      console.log("[EventService] Received event:", data.type, data)
      this.handleEvent(data)
    })

    // Listen for streaming events
    this.socket.on("stream_event", (data: any) => {
      console.log("[EventService] Received stream event:", data.event_type, data)
      // Treat stream_event as a special event type for subscribers
      this.handleStreamEvent(data)
    })

    // Ping/pong for keepalive
    this.socket.on("pong", (data) => {
      console.log("[EventService] Pong received:", data)
    })
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect(): void {
    if (this.socket) {
      console.log("[EventService] Disconnecting...")
      this.socket.disconnect()
      this.socket = null
      this.isConnecting = false
      this.subscriptions.clear()
      this.activeRooms.clear()
      this.updateStatus("disconnected")
    }
  }

  /**
   * Subscribe to specific event type or all events
   * @param eventType - Event type to subscribe to, or "*" for all events
   * @param handler - Callback function to handle the event
   * @returns Subscription ID (use to unsubscribe)
   */
  subscribe(eventType: string | "*", handler: EventHandler): string {
    const id = `${eventType}_${Date.now()}_${Math.random()}`
    this.subscriptions.set(id, { eventType, handler, id })
    console.log(`[EventService] Subscribed to ${eventType}, subscription ID:`, id)
    return id
  }

  /**
   * Unsubscribe from an event
   * @param subscriptionId - ID returned from subscribe()
   */
  unsubscribe(subscriptionId: string): void {
    const subscription = this.subscriptions.get(subscriptionId)
    if (subscription) {
      this.subscriptions.delete(subscriptionId)
      console.log(`[EventService] Unsubscribed from ${subscription.eventType}`)
    }
  }

  /**
   * Subscribe to a specific room
   * @param room - Room name to subscribe to
   */
  async subscribeToRoom(room: string): Promise<void> {
    // Track room regardless of connection state so it gets re-subscribed on reconnect
    this.activeRooms.add(room)

    if (!this.socket) {
      console.warn("[EventService] Cannot subscribe to room: not connected (will subscribe on reconnect)")
      return
    }

    return new Promise((resolve, reject) => {
      this.socket?.emit("subscribe", { room }, (response: any) => {
        if (response?.status === "success") {
          console.log(`[EventService] Subscribed to room: ${room}`)
          resolve()
        } else {
          console.error(`[EventService] Failed to subscribe to room: ${room}`, response)
          reject(new Error(response?.message || "Failed to subscribe"))
        }
      })
    })
  }

  /**
   * Unsubscribe from a specific room
   * @param room - Room name to unsubscribe from
   */
  async unsubscribeFromRoom(room: string): Promise<void> {
    this.activeRooms.delete(room)

    if (!this.socket) {
      console.warn("[EventService] Cannot unsubscribe from room: not connected")
      return
    }

    return new Promise((resolve, reject) => {
      this.socket?.emit("unsubscribe", { room }, (response: any) => {
        if (response?.status === "success") {
          console.log(`[EventService] Unsubscribed from room: ${room}`)
          resolve()
        } else {
          console.error(`[EventService] Failed to unsubscribe from room: ${room}`, response)
          reject(new Error(response?.message || "Failed to unsubscribe"))
        }
      })
    })
  }

  /**
   * Send a ping to check connection
   */
  ping(): void {
    if (this.socket) {
      this.socket.emit("ping")
    }
  }

  /**
   * Send agent usage intent event to backend
   * This signals that the user intends to use a specific agent environment
   * and triggers activation if the environment is suspended
   *
   * @param environmentId - Environment UUID
   * @returns Promise resolving to backend response
   */
  async sendAgentUsageIntent(environmentId: string): Promise<any> {
    if (!this.socket) {
      console.warn("[EventService] Cannot send agent_usage_intent: not connected")
      return { status: "error", message: "Not connected" }
    }

    return new Promise((resolve, reject) => {
      this.socket?.emit("agent_usage_intent", { environment_id: environmentId }, (response: any) => {
        if (response?.status === "error") {
          console.error(`[EventService] agent_usage_intent error:`, response)
          reject(new Error(response.message || "Failed to send usage intent"))
        } else {
          console.log(`[EventService] agent_usage_intent sent for environment ${environmentId}, response:`, response)
          resolve(response)
        }
      })
    })
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.socket?.connected || false
  }

  /**
   * Get socket ID
   */
  getSocketId(): string | undefined {
    return this.socket?.id
  }

  /**
   * Get current connection status
   */
  getStatus(): ConnectionStatus {
    return this.currentStatus
  }

  /**
   * Subscribe to connection status changes
   * @param listener - Callback function to handle status changes
   * @returns Unsubscribe function
   */
  onStatusChange(listener: ConnectionStatusListener): () => void {
    this.statusListeners.add(listener)
    // Immediately call with current status
    listener(this.currentStatus)

    // Return unsubscribe function
    return () => {
      this.statusListeners.delete(listener)
    }
  }

  /**
   * Update connection status and notify listeners
   */
  private updateStatus(status: ConnectionStatus): void {
    if (this.currentStatus !== status) {
      this.currentStatus = status
      console.log(`[EventService] Status changed to: ${status}`)
      this.statusListeners.forEach((listener) => {
        try {
          listener(status)
        } catch (error) {
          console.error("[EventService] Error in status listener:", error)
        }
      })
    }
  }

  /**
   * Handle incoming event and notify subscribers
   */
  private handleEvent(event: EventData): void {
    // Notify subscribers who are listening to this specific event type
    this.subscriptions.forEach((subscription) => {
      if (subscription.eventType === event.type || subscription.eventType === "*") {
        try {
          subscription.handler(event)
        } catch (error) {
          console.error(`[EventService] Error in event handler for ${event.type}:`, error)
        }
      }
    })
  }

  /**
   * Handle incoming stream event and notify subscribers
   */
  private handleStreamEvent(streamEvent: any): void {
    // Notify subscribers who are listening to "stream_event"
    this.subscriptions.forEach((subscription) => {
      if (subscription.eventType === "stream_event" || subscription.eventType === "*") {
        try {
          subscription.handler(streamEvent)
        } catch (error) {
          console.error(`[EventService] Error in stream event handler:`, error)
        }
      }
    })
  }
}

// Export singleton instance
export const eventService = new EventServiceClass()

File: backend/app/services/events/event_service.py
"""EventService for managing WebSocket-based real-time events."""

import logging
import asyncio
from datetime import datetime, UTC
from typing import Any, Callable, Awaitable
from uuid import UUID
import concurrent.futures

from app.core.db import create_session
from app.models.events.event import EventPublic, EventBroadcast
from app.services.events import socket_auth
from app.services.events.socketio_connector import socketio_connector
from app.utils import create_task_with_error_logging

logger = logging.getLogger(__name__)

# Type alias for event handler functions
EventHandler = Callable[[dict[str, Any]], Awaitable[None]]


class EventService:
    """Service for managing real-time events via WebSocket."""

    def __init__(self):
        """Initialize the event service with a Socket.IO async server."""
        # Use the injectable connector's sio for handler registration / rooms
        self.sio = socketio_connector.sio

        # Track active connections: {sid: ConnectionInfo}
        self.connections: dict[str, dict[str, Any]] = {}

        # Thread pool for background tasks (to avoid blocking event loop)
        self.executor = concurrent.futures.ThreadPoolExecutor(
            max_workers=4,
            thread_name_prefix="event_service_bg"
        )

        # Backend event handlers registry: {event_type: [handler_functions]}
        self._backend_handlers: dict[str, list[EventHandler]] = {}

        # Register event handlers
        self._register_handlers()

    def _register_handlers(self):
        """Register Socket.IO event handlers."""

        @self.sio.event
        async def connect(sid, environ, auth):
            """Handle client connection.

            The socket app is mounted publicly at ``/ws``, so this is the only
            authentication gate on the event stream. Identity comes from the
            signed ``auth["token"]`` and nothing else — a client-supplied user
            id is ignored. See ``events/socket_auth.py``.
            """
            logger.info(f"Client connecting: {sid}")

            with create_session() as db_session:
                principal = socket_auth.resolve_principal(db_session, auth)

            if principal is None:
                logger.warning(f"Connection {sid} rejected: no valid token in auth")
                return False  # Reject connection

            # Store connection info
            self.connections[sid] = {
                "sid": sid,
                "principal": principal,
                "user_id": principal.id,
                "connected_at": datetime.now(UTC),
                "rooms": [],
            }

            # Join the principal's own room
            await self.sio.enter_room(sid, principal.room)
            self.connections[sid]["rooms"].append(principal.room)

            logger.info(
                f"Client {sid} connected as {principal.kind} {principal.id}, "
                f"joined room: {principal.room}"
            )
            return True

        @self.sio.event
        async def disconnect(sid):
            """Handle client disconnection."""
            if sid in self.connections:
                user_id = self.connections[sid]["user_id"]
                logger.info(f"Client {sid} disconnected (user: {user_id})")
                del self.connections[sid]
            else:
                logger.info(f"Client {sid} disconnected (unknown)")

        @self.sio.event
        async def subscribe(sid, data):
            """Handle subscription to specific event types or rooms.

            Args:
                data: Dict with 'room' or 'event_type' to subscribe to
            """
            if sid not in self.connections:
                logger.warning(f"Subscribe request from unknown connection: {sid}")
                return {"status": "error", "message": "Not authenticated"}

            room = data.get("room") if isinstance(data, dict) else None
            if not room:
                return {"status": "error", "message": "No room specified"}

            # A room is joined only if it belongs to this connection's
            # principal. Without this, any authenticated client could enter any
            # session's stream room by name.
            principal = self.connections[sid]["principal"]
            with create_session() as db_session:
                allowed = socket_auth.can_join_room(db_session, principal, room)

            if not allowed:
                logger.warning(
                    f"Client {sid} ({principal.kind} {principal.id}) refused "
                    f"room: {room}"
                )
                return {"status": "error", "message": "Not authorized for room"}

            await self.sio.enter_room(sid, room)
            self.connections[sid]["rooms"].append(room)
            logger.info(f"Client {sid} subscribed to room: {room}")
            return {"status": "success", "room": room}

        @self.sio.event
        async def unsubscribe(sid, data):
            """Handle unsubscription from specific rooms.

            Args:
                data: Dict with 'room' to unsubscribe from
            """
            if sid not in self.connections:
                logger.warning(f"Unsubscribe request from unknown connection: {sid}")
                return {"status": "error", "message": "Not authenticated"}

            room = data.get("room")
            if room:
                await self.sio.leave_room(sid, room)
                if room in self.connections[sid]["rooms"]:
                    self.connections[sid]["rooms"].remove(room)
                logger.info(f"Client {sid} unsubscribed from room: {room}")
                return {"status": "success", "room": room}

            return {"status": "error", "message": "No room specified"}

        @self.sio.event
        async def ping(sid):
            """Handle ping from client (for keepalive)."""
            return {"status": "pong", "timestamp": datetime.now(UTC).isoformat()}

        @self.sio.event
        async def agent_usage_intent(sid, data):
            """
            Handle agent usage intent event from frontend.

            This event is sent when the user shows intention to use an agent:
            - Opening a session with that agent
            - Clicking on the agent in the dashboard
            - Navigating to agent's page

            Args:
                data: Dict with 'environment_id' and optionally 'agent_id'

            Returns:
                Status dict indicating if activation was triggered
            """
            if sid not in self.connections:
                logger.warning(f"agent_usage_intent from unknown connection: {sid}")
                return {"status": "error", "message": "Not authenticated"}

            principal = self.connections[sid]["principal"]
            if principal.kind != "user":
                # A webapp viewer holds a share id, not a user id — handing it
                # to a user-scoped service would be a category error.
                logger.warning(
                    f"agent_usage_intent refused for {principal.kind} {principal.id}"
                )
                return {"status": "error", "message": "Not authorized"}

            user_id = principal.id
            environment_id = data.get("environment_id")

            if not environment_id:
                logger.warning(f"agent_usage_intent without environment_id from user {user_id}")
                return {"status": "error", "message": "environment_id required"}

            try:
                # Delegate to the shared service function. The WS handler trusts
                # the socket's authenticated user_id and does no extra access
                # control (the REST route enforces ownership; see usage_intent.py).
                from app.core.db import engine as db_engine
                from sqlmodel import Session as DBSession
                from app.services.environments.usage_intent import register_usage_intent

                with DBSession(db_engine) as session:
                    return register_usage_intent(
                        db_session=session,
                        user_id=user_id,
                        environment_id=UUID(environment_id),
                    )

            except ValueError as e:
                logger.warning(f"agent_usage_intent rejected: {e}")
                return {"status": "error", "message": str(e)}
            except Exception as e:
                logger.error(f"Error handling agent_usage_intent: {e}", exc_info=True)
                return {"status": "error", "message": str(e)}

    def register_handler(self, event_type: str, handler: EventHandler):
        """Register a backend handler for a specific event type.

        This allows backend services to react to events without using WebSockets.

        Args:
            event_type: Event type to listen for (e.g., 'stream_completed')
            handler: Async function that accepts event data dict
        """
        if event_type not in self._backend_handlers:
            self._backend_handlers[event_type] = []
        self._backend_handlers[event_type].append(handler)
        logger.info(f"Registered backend handler for event type: {event_type}")

    async def _call_backend_handlers(
        self, event_type: str, event_data: dict[str, Any], *,
        deferred_handlers: list[tuple[EventHandler, dict[str, Any]]] | None = None,
    ):
        """Call all registered backend handlers for an event type.

        Args:
            event_type: Event type
            event_data: Full event data including type, model_id, meta, etc.
        """
        handlers = self._backend_handlers.get(event_type, [])
        if not handlers:
            return

        logger.debug(f"Calling {len(handlers)} backend handler(s) for event type: {event_type}")

        if deferred_handlers is not None:
            # A channel processor drains terminal callbacks after its relay is
            # stopped, while still holding the session lock and current address.
            deferred_handlers.extend((handler, event_data) for handler in handlers)
            return

        # Call all handlers in background tasks (non-blocking)
        for i, handler in enumerate(handlers):
            try:
                # Create task to run handler without awaiting
                # Use error logging wrapper to prevent silent failures and premature cancellation
                create_task_with_error_logging(
                    handler(event_data),
                    task_name=f"event_handler_{event_type}_{i}"
                )
            except Exception as e:
                logger.error(f"Error calling backend handler for {event_type}: {e}", exc_info=True)

    async def emit_event(
        self,
        event_type: str,
        model_id: UUID | None = None,
        text_content: str | None = None,
        meta: dict[str, Any] | None = None,
        user_id: UUID | None = None,
        room: str | None = None,
        *,
        deferred_handlers: list[tuple[EventHandler, dict[str, Any]]] | None = None,
    ):
        """Emit an event to connected clients and backend handlers.

        Args:
            event_type: Type of event (e.g., 'session_updated')
            model_id: ID of the related model
            text_content: Optional notification text
            meta: Additional metadata
            user_id: Target specific user (will send to user_{user_id} room)
            room: Target specific room (alternative to user_id)
            deferred_handlers: Internal channel terminal callbacks, drained by
                the session processor after its relay stops and before its lock releases.
        """
        event = EventPublic(
            type=event_type,
            model_id=model_id,
            text_content=text_content,
            meta=meta or {},
            user_id=user_id,
            timestamp=datetime.now(UTC),
        )

        event_data = event.model_dump(mode="json")

        # Ordinary events schedule handlers immediately. A channel terminal
        # event can defer them until its processor has stopped the live relay.
        if deferred_handlers is None:
            await self._call_backend_handlers(event_type, event_data)
        else:
            await self._call_backend_handlers(event_type, event_data, deferred_handlers=deferred_handlers)

        # Determine target room
        target_room = room
        if user_id and not target_room:
            target_room = f"user_{user_id}"

        if target_room:
            # Send to specific room
            logger.info(f"Emitting event {event_type} to room {target_room}")
            await socketio_connector.emit("event", event_data, room=target_room)
        else:
            # Broadcast to all connected clients
            logger.info(f"Broadcasting event {event_type} to all clients")
            await socketio_connector.emit("event", event_data)

    async def broadcast_event(self, broadcast: EventBroadcast):
        """Broadcast an event using EventBroadcast model.

        Args:
            broadcast: EventBroadcast model with all event details
        """
        await self.emit_event(
            event_type=broadcast.type,
            model_id=broadcast.model_id,
            text_content=broadcast.text_content,
            meta=broadcast.meta,
            user_id=broadcast.user_id,
            room=broadcast.room,
        )

    def _user_connections(self) -> list[dict[str, Any]]:
        """Connections held by a real user.

        A public webapp viewer also holds a connection, but its principal id is
        an ``AgentWebappShare`` id, not a user id — so it must not count towards
        "is this user online".
        """
        return [
            conn
            for conn in self.connections.values()
            if conn["principal"].kind == "user"
        ]

    def get_connected_users(self) -> list[UUID]:
        """Get list of currently connected user IDs."""
        return list({conn["user_id"] for conn in self._user_connections()})

    def is_user_online(self, user_id: UUID) -> bool:
        """
        Check if a specific user is online (has active WebSocket connection).

        Args:
            user_id: User UUID

        Returns:
            True if user has at least one active connection
        """
        return any(conn["user_id"] == user_id for conn in self._user_connections())

    def is_user_connected(self, user_id: UUID) -> bool:
        """Check if a specific user is connected (alias for is_user_online)."""
        return self.is_user_online(user_id)

    def get_connection_count(self) -> int:
        """Get total number of active connections."""
        return len(self.connections)

    async def emit_stream_event(
        self,
        session_id: UUID,
        event_type: str,
        event_data: dict[str, Any],
    ):
        """
        Emit a streaming event to a session-specific room.

        Args:
            session_id: Session UUID
            event_type: Type of streaming event (assistant, tool, etc.)
            event_data: Full event data including content, metadata
        """
        room = f"session_{session_id}_stream"

        # Emit to session-specific streaming room
        await socketio_connector.emit(
            "stream_event",
            {
                "session_id": str(session_id),
                "event_type": event_type,
                "data": event_data,
                "timestamp": datetime.now(UTC).isoformat()
            },
            room=room
        )

        logger.debug(f"Emitted stream event {event_type} to room {room}")

    def _activate_environment_sync(self, environment_id: str, agent_id: str):
        """
        Synchronous background task to activate a suspended environment.
        Runs in a separate thread to avoid blocking the event loop.

        Args:
            environment_id: Environment UUID as string
            agent_id: Agent UUID as string
        """
        try:
            # Import here to avoid circular dependencies
            from app.core.db import engine as db_engine
            from sqlmodel import Session as DBSession
            from app.models.environments.environment import AgentEnvironment
            from app.models.agents.agent import Agent
            from app.services.environments.environment_lifecycle import EnvironmentLifecycleManager

            # Use fresh DB session for background task
            with DBSession(db_engine) as session:
                environment = session.get(AgentEnvironment, UUID(environment_id))
                agent = session.get(Agent, UUID(agent_id))

                if not environment or not agent:
                    logger.error(f"Environment or agent not found: env={environment_id}, agent={agent_id}")
                    return

                # Activate the environment using asyncio.run for async operations
                lifecycle_manager = EnvironmentLifecycleManager()
                asyncio.run(
                    lifecycle_manager.activate_suspended_environment(
                        db_session=session,
                        environment=environment,
                        agent=agent,
                        emit_events=True
                    )
                )

                logger.info(f"Background activation completed for environment {environment_id}")

        except Exception as e:
            logger.error(f"Background activation failed for environment {environment_id}: {e}", exc_info=True)

    def shutdown(self):
        """Shutdown the event service and cleanup resources."""
        logger.info("Shutting down EventService executor...")
        self.executor.shutdown(wait=True, cancel_futures=True)
        self._backend_handlers.clear()
        logger.info("EventService executor shut down")

    def get_asgi_app(self):
        """Get the ASGI app for Socket.IO."""
        return socketio_connector.get_asgi_app()


# Global event service instance
event_service = EventService()
